pkg/cmd/repo/sync/sync.go

```diff
@@ -0,0 +1,325 @@
+		opts.IO.StopProgressIndicator()
+		if err != nil {
+			return err
+		}
+		if srcRepo == nil {
+			return fmt.Errorf("can't determine source repo for %s because repo is not fork", ghrepo.FullName(destRepo))
+		}
+	} else {
+		srcRepo, err = ghrepo.FromFullName(opts.SrcArg)
+		if err != nil {
+			return err
+		}
+	}
+
+	if destRepo.RepoHost() != srcRepo.RepoHost() {
+		return fmt.Errorf("can't sync repos from different hosts")
+	}
+
+	if opts.Branch == "" {
+		opts.IO.StartProgressIndicator()
+		opts.Branch, err = api.RepoDefaultBranch(apiClient, srcRepo)
+		opts.IO.StopProgressIndicator()
+		if err != nil {
+			return err
+		}
+	}
+
+	opts.IO.StartProgressIndicator()
+	err = executeRemoteRepoSync(apiClient, destRepo, srcRepo, opts)
+	opts.IO.StopProgressIndicator()
+	if err != nil {
+		if errors.Is(err, divergingError) {
+			return fmt.Errorf("can't sync because there are diverging changes, you can use `--force` to overwrite the changes")
+		}
+		return err
+	}
+
+	if opts.IO.IsStdoutTTY() {
+		cs := opts.IO.ColorScheme()
+		srcStr := fmt.Sprintf("%s:%s", ghrepo.FullName(srcRepo), opts.Branch)
+		destStr := fmt.Sprintf("%s:%s", ghrepo.FullName(destRepo), opts.Branch)
+		success := fmt.Sprintf("Synced %s from %s\n", cs.Bold(destStr), cs.Bold(srcStr))
+		fmt.Fprintf(opts.IO.Out, "%s %s", cs.SuccessIcon(), success)
+	}
+
+	return nil
+}
+
+var divergingError = errors.New("diverging changes")
+var mismatchRemotesError = errors.New("branch remote does not match specified source")
+
+func executeLocalRepoSync(srcRepo ghrepo.Interface, remote string, opts *SyncOptions) error {
+	git := opts.Git
+	branch := opts.Branch
+
+	err := git.Fetch([]string{remote, fmt.Sprintf("+refs/heads/%s", branch)})
+	if err != nil {
+		return err
+	}
+
+	hasLocalBranch := git.HasLocalBranch([]string{branch})
+	if hasLocalBranch {
+		branchRemote, err := git.BranchRemote(branch)
+		if err != nil {
+			return err
+		}
+		if branchRemote != remote {
+			return mismatchRemotesError
+		}
+
+		fastForward, err := git.IsAncestor([]string{branch, fmt.Sprintf("%s/%s", remote, branch)})
+		if err != nil {
+			return err
+		}
+
+		if !fastForward && !opts.Force {
+			return divergingError
+		}
+	}
+
+	startBranch, err := git.CurrentBranch()
+	if err != nil {
+		return err
+	}
+	if startBranch != branch {
+		if hasLocalBranch {
+			err = git.Checkout([]string{branch})
+			if err != nil {
+				return err
+			}
+		} else {
+			err = git.Checkout([]string{"--track", fmt.Sprintf("%s/%s", remote, branch)})
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if hasLocalBranch {
+		if opts.Force {
+			err = git.Reset([]string{"--hard", fmt.Sprintf("refs/remotes/%s/%s", remote, branch)})
+			if err != nil {
+				return err
+			}
+		} else {
+			err = git.Merge([]string{"--ff-only", fmt.Sprintf("refs/remotes/%s/%s", remote, branch)})
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if startBranch != branch {
+		err = git.Checkout([]string{startBranch})
```

Comment:
Not sure I understand the question. The first `git.Checkout` is to checkout the branch to sync and then this `git.Checkout` is to switch back to the initial branch the user was on. I didn't want to assume that the user wanted to switch to the branch that was being sync'ed.

Reply:
Ah I've missed that detail. We do not have to switch branches just to update a non-current branch to another ref: there is `git update-ref` for that! The benefit of not switching branches unnecessarily is that the filesystem needs fewer writes (since the working copy is not being updated) and git reflog is cleaner, which benefits tooling. I'll open a PR to address this.